x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmTestCase.java

```diff
@@ -430,20 +434,41 @@ protected void doMultipleAuthcAuthzAndVerifySuccess(
                     equalTo(Map.of("jwt_token_type", JwtRealmInspector.getTokenType(jwtRealm).value()))
                 );
             }
+            // if the cache is enabled ensure the cache is used and does not change for the provided jwt
+            if (jwtRealm.getJwtCache() != null) {
+                Cache<BytesArray, JwtRealm.ExpiringUser> cache = jwtRealm.getJwtCache();
+                if (firstCacheKeyFound == null) {
+                    assertNotNull("could not find cache keys", cache.keys());
+                    firstCacheKeyFound = cache.keys().iterator().next();
+                }
+                jwtAuthenticationToken.clearCredentials(); // simulates the realm's context closing which clears the credential
+                boolean foundInCache = false;
+                for (BytesArray key : cache.keys()) {
+                    logger.trace("cache key: " + HexFormat.of().formatHex(key.array()));
+                    if (key.equals(firstCacheKeyFound)) {
+                        foundInCache = true;
+                    }
+                    assertFalse(
+                        "cache key should not be nulled out",
+                        IntStream.range(0, key.array().length).map(idx -> key.array()[idx]).allMatch(b -> b == 0)
+                    );
+                }
+                assertTrue("cache key was not found in cache", foundInCache);
```

Comment:
I understand that the test runs authn multiple times for the same JWT token, and that here you're asserting that, for the subsequent authn attempts (with the same JWT), the cache still contains at least the entry populated during the first authn.

Sounds to me like a rather weak guarantee. A strong guarantee would check that the cache contains exactly one entry and that the `jwtAuthenticator` is only invoked once (because the cache filters out subsequent invokations).

Do you think it's easily feasible and worth it to strenghten the assertion  here? (I realize that the `JwtAuthenticator` is not exposed as a dependency in the `JwtRealm` currently).

Reply:
you're right it is a fairly weak guarantee. however, since this can be called from any test any number of times we are limited here in what we can actually assert. To compensate, I beefed up the single test to account for stronger guarantees for how many items are expected be in the cache https://github.com/elastic/elasticsearch/pull/101799/commits/89519b1615856453b687421314b810c24aa8c3f6.